### PR TITLE
[NestedTensor] Integrate the softmax operator along the jagged dimension into NestedTensor

### PR DESCRIPTION
**Pull request by jananisriram**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131519
* __->__ #131518

Modify the existing `softmax` operator in PyTorch, invoked by `torch.softmax`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the aten padding operator, enables PyTorch users to invoke `torch.softmax` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` nested tensor.

Write unit tests based on the `sum` and `mean` jagged operators to verify the accuracy of the ragged reduction implementation for `torch.softmax`. Add unit tests to verify error handling for unsupported features in `NestedTensor` `torch.softmax`.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. In addition, the `softmax` operator is required to take in as input an integer for the reduction dimension `dim`, requiring new unit tests heavily inspired by the `sum` and `mean` jagged operator unit tests. `Softmax` also allows for reducing along the batch dimension.

### CODE REVIEW

**Comment by jananisriram** [comment]
@jananisriram has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59951075).

**Comment by jananisriram** [comment]
@jananisriram has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59951075).

File: test/test_nestedtensor.py
Comment (davidberard98) [inline test/test_nestedtensor.py:233]: curious why you need the `_` ?

File: test/test_nestedtensor.py
Comment (davidberard98) [inline test/test_nestedtensor.py:4159]: nice solution!

File: torch/nested/_internal/ops.py
Comment (davidberard98) [inline torch/nested/_internal/ops.py:540]: I don't think this is correct. We should probably error if we try to reduce on dim=0.

To check this, I think you can do something like this:
```
x = torch.randn(4, 4, 4)
nt = torch.nested.as_nested_tensor(x, layout=torch.jagged)
dense_softmax = torch.nn.functional.softmax(x, dim=0)
nt_softmax = torch.nn.functional.softmax(nt, dim=0)
assert torch.allclose(dense_softmax, nt_softmax.values())
```

**Review by davidberard98 — CHANGES_REQUESTED** [review]

File: test/test_nestedtensor.py
Comment (jananisriram) [reply test/test_nestedtensor.py:233]: depending on the op, `layout.__name__.split(".")[0]` returns `_softmax` rather than `softmax`, so wanted to remove the `_`!

File: torch/nested/_internal/ops.py
Comment (jananisriram) [reply torch/nested/_internal/ops.py:540]: noted, and added a test/assert for this!

File: torch/nested/_internal/ops.py
Comment (davidberard98) [inline torch/nested/_internal/ops.py:39]: nit: since you aren't using allow_batch_dim, you could probably remove the changes to this file

File: torch/nested/_internal/ops.py
Comment (davidberard98) [inline torch/nested/_internal/ops.py:512]: nit: why is `dim` changed to `any?`? is dim optional?

**Review by davidberard98 — APPROVED** [review]
a few nits, otherwise LGTM! (also pending merge tests etc.)

File: torch/nested/_internal/ops.py
Comment (jananisriram) [reply torch/nested/_internal/ops.py:512]: that's a good point; i was getting some errors with `dim` being non-optional for `sum` and `mean` so wanted to prevent that for `softmax`, but it should definitely require a `dim`!

File: torch/nested/_internal/ops.py
Comment (jananisriram) [reply torch/nested/_internal/ops.py:39]: thanks! removed the condition for `allow_batch_dim`

**Review by ZainRizvi — CHANGES_REQUESTED** [review]
Note: Please do not merge this PR.  Create a new codev diff and only use the codev tools to push the changes to github.  Alternatively, you can create the new PR directly on github and the pytorch diff train will take care of get it merged internally

**Pull request closed without merging**
